### PR TITLE
[WIP]Implement attribute assortativity app and numeric assortativity app

### PR DESCRIPTION
**Pull request by Ningsir**
## What do these changes do?

Implement attribute assortativity coefficient app and numeric assortativity coefficient.

- [x] implement C++ PIE app with analytical engineAPI;
-  [x] check the correctness with NetworkX
-  [x] check the scalability and performance in large graph, like com-friendster dataset
-  [x] add to python built-in algorithms and add related test case

## Related issue number
#453

## Experiment

**attribute assortativity coefficient**：

| Time           | GraphScope | NetworkX  |
| -------------- | ---------- | --------- |
| Twitter        | 0.58  S    | 1.32 S    |
| datagen-7_5-fb | 110  S     | 1028.03 S |
| datagen-7_7-zf | 130 S      | 1048.43 S |
| datagen-8_0-fb | 198.85 S   | 1899.76 S |
|                |            |           |
| Memory Usage   |            |           |
| Twitter        | 0.64 G     | 0.71 G    |
| datagen-7_5-fb | 9.6  G     | 16.35 G   |
| datagen-7_7-zf | 27.2 G     | 26.01 G   |
| datagen-8_0-fb | 30.72 G    | 28.41 G   |

**numeric assortativity coefficient**：

| Time           | GraphScope | NetworkX |
| -------------- | ---------- | -------- |
| Twitter        | 0.58 S     | 1.51 S   |
| datagen-7_5-fb | 101.52 S   | 188.64 S |
| datagen-7_7-zf | 134.13 S   | 110.27 S |
| datagen-8_0-fb | 219.91 S   | OOM      |
|                |            |          |
| Memory Usage   |            |          |
| Twitter        | 0.96 G     | 0.71 G   |
| datagen-7_5-fb | 11.20 G    | 17.75 G  |
| datagen-7_7-zf | 25 G       | 27.56 G  |
| datagen-8_0-fb | 31 G       | OOM      |

### CODE REVIEW

**Comment by yecol** [comment]
@Ningsir Thanks for your commit!
Please assign the related issue in the abstract.

File: analytical_engine/apps/assortativity/attribute/attribute_assortativity.h [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+               message_manager_t& messages) {
+    if (!ctx.merge_stage) {
+      vdata_t source_data;
+      vertex_t u;
+      while (messages.GetMessage(frag, u, source_data)) {
+        vdata_t target_data = frag.GetData(u);
+        attributeMixingCount(source_data, target_data, ctx);
+      }
+      ctx.merge_stage = true;
+      // send message to work 0
```
Comment (acezen) [inline analytical_engine/apps/assortativity/attribute/attribute_assortativity.h:73]: work->worker

File: analytical_engine/apps/assortativity/attribute/attribute_assortativity.h [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+      if (frag.fid() != 0) {
+        messages.SendToFragment(0, ctx.attribute_mixing_map);
+      }
+      messages.ForceContinue();
+    } else {
+      // merge in work 0
```
Comment (acezen) [inline analytical_engine/apps/assortativity/attribute/attribute_assortativity.h:79]: work -> worker

**Review by acezen — CHANGES_REQUESTED** [review]
why we nedd `numeric_assortativity.h`, seems we can reuse the `attribute_assortativity.h`, please double check?

File: analytical_engine/apps/assortativity/attribute/attribute_assortativity.h (hunk 1 above)
Comment (Ningsir) [reply analytical_engine/apps/assortativity/attribute/attribute_assortativity.h:73]: the mixing matrix of attribute asssortativity app is same with numeric assortativity, but their formula for getting result is different.

**Comment by acezen** [comment]
Can we implement the two algorithms as one app ? 
we can send a parameter to check what kind of matrix computation we need

File: python/graphscope/nx/algorithms/builtin.py [hunk 3]
```diff
@@ -1045,6 +1048,97 @@ def edge_boundary(G, nbunch1, nbunch2=None):
...
+    where M is the joint probability distribution (mixing matrix)
+    of the specified attribute.
+
+    References
+    ----------
+    .. [1] M. E. J. Newman, Mixing patterns in networks,
+       Physical Review E, 67 026126, 2003
+    """
+
+    ctx = AppAssets(algo="attribute_assortativity_coefficient", context="tensor")(
```
Comment (acezen) [inline python/graphscope/nx/algorithms/builtin.py:1091]: like [degree_assortativity_coefficient](https://github.com/alibaba/GraphScope/blob/13b6b4f87874fb0c57a5d3d2dbfe703dd6027b9c/python/graphscope/nx/algorithms/builtin.py#L950), here we shoule call `graphscope.attribute_assortativity_coefficient`

File: python/graphscope/nx/algorithms/builtin.py [hunk 4]
```diff
@@ -1045,6 +1048,97 @@ def edge_boundary(G, nbunch1, nbunch2=None):
...
+    This computes Eq. (21) in Ref. [1]_ , for the mixing matrix
+    of the specified attribute.
+
+    References
+    ----------
+    .. [1] M. E. J. Newman, Mixing patterns in networks
+           Physical Review E, 67 026126, 2003
+    """
+
+    ctx = AppAssets(algo="attribute_assortativity_coefficient", context="tensor")(
```
Comment (acezen) [inline python/graphscope/nx/algorithms/builtin.py:1136]: Ditto

**Review by acezen — COMMENTED** [review]
add a p2p test case for `attribute assortativity` and `numeric assortativity `like [degree_assortativity_coefficient](https://github.com/alibaba/GraphScope/blob/13b6b4f87874fb0c57a5d3d2dbfe703dd6027b9c/python/graphscope/nx/tests/test_ctx_builtin.py#L250)

**Review by acezen — APPROVED** [review]
Overall LGTM.

**Pull request merged**